kep/260-network-aware-scheduling/README.md

```diff
@@ -0,0 +1,1511 @@
+so network latency and bandwidth between nodes are considered while making scheduling decisions.
+Data centers with fat-tree topology or cluster topology can benefit from our network-aware framework, 
+as network conditions (i.e., network latency, available bandwidth) between nodes can vary according to their 
+locations in the infrastructure. 
+
+<p align="center"><img src="figs/cluster.png" title="Cluster Topology" width="600" class="center"/></p>
+
+<p align="center"><img src="figs/data_center.png" title="DC Topology" width="600" class="center"/></p>
+
+Deploying microservices on different sets of nodes will impact the application's response time. 
+For specific applications, latency and bandwidth requirements can be critical.  
+For example, in a [Redis cluster](https://redis.io/topics/cluster-tutorial), 
+master nodes need to synchronize data with slave nodes frequently. Namely, there are dependencies between the 
+masters and the slaves. High latency or low bandwidth between masters and slaves can lead to slow CRUD operations.
+
+<p align="center"><img src="figs/redis.png" title="Redis app" width="600" class="center"/></p>
+
+### 2 - Cloud2Edge application running on a multi-region geo-distributed cluster.
+
+Multi-region Geo-distributed scenarios benefit the most from our framework and network-aware plugins. 
+
+<p align="center"><img src="figs/multi_region.png" title="MultiRegion Topology" width="600" class="center"/></p>
+
+High latency is a big concern in these topologies, especially for IoT applications 
+(e.g., [Eclipse Hono](https://github.com/eclipse/hono), [Eclipse Cloud2Edge](https://www.eclipse.org/packages/packages/cloud2edge/)). 
+For example, in the Cloud2Edge platform, there are several dependencies among the several APIs and MQTT brokers where devices connect to:
+
+<p align="center"><img src="figs/cloud2edge.png" title="Cloud2Edge" width="600" class="center"/></p>
+
+# Proposal - Design & Implementation details
+
+## Overview of the System Design
+
+The proposal introduces two [Custom Resources (CRs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) 
+defined as Custom Resource Definitions (CRDs), **AppGroup CRD** and **NetworkTopology CRD**, to maintain 
+both the service topology information from application microservice dependencies and the infrastructure 
+network topology where network weights are established between regions and zones in the cluster. 
+Thus, both application and infrastructure network topology are considered during scheduling. 
+
+The proposal also presents a bandwidth resource component (DaemonSet) to advertise bandwidth resources 
+as **extended resources** to allow already available filter/scoring plugins (e.g., `PodFitsResources`, 
+`BalancedAlocation`) to consider bandwidth allocation.
+
+As an initial design, we plan to implement two plugins:  
+- A **QueueSort** function named `TopologicalSort` where pods are sorted based on their established dependencies.
```

Comment:
Thanks @Huang-Wei for your comment since it raises an important question. 

The idea is to propose an additional QueueSort plugin to find the preferred order in which pods belonging to the same AppGroup should be scheduled. 

We noticed in our experiments that the order from which the pods are scheduled is extremely important and affects the overall application end-to-end latency. The proposed TopologicalSort plugin calculates the preferred order for a certain AppGroup based on the dependencies described in the AppGroup CR. 

In the design, we favor topology sorting instead of priority sorting, but users can still opt to enable the default QueueSort plugin if priorities are important for their use case. In addition, we consider as default the less function of the QoS plugin, when pods do not belong to an AppGroup. 

At first sight, the idea of the Prefilter plugin could lead to deadlock situations since pods might not be scheduled because any of his dependencies are missing. So, eventually not getting scheduled in the cluster. 

We updated the KEP with further details about the QueueSort plugin. We can also share some early results on topological sorting showing the differences concerning the application end-to-end response time.

Reply:
Thanks for the explanations. Just keep in mind that if two plugins both implements QueueSort, they cannot be enabled together.